src/core/ref-counted-object.js

```diff
@@ -1,8 +1,11 @@
 // base class for reference counted objects
 class RefCountedObject {
-    constructor() {
+    constructor(autoDestroy = true) {
```

Comment:
Would an option be to update the element system to respect a destroy on the ref count reaching 0?

Reply:
I considered this, but auto destroying meshes change could still cause errors in existing projects, that internally keep meshes around for some reason without incReferencing them - they'd end up keeping mesh that has been auto destroyed by the engine.

Reply:
So why ref count meshes at all? Isn't the whole point to automate destruction?

Reply:
Well the way it's been done (and I reverted to that original solution) is that meshes are refCounted, but what happens with them when ref count reaches zero depends on the owner of the mesh - and they don't always get destroyed, but they get reused.

We can change it and fix the engine to be consistent, but what about external code? I'd be happy with this option if we're happy to take the risk.

Reply:
I would probably prefer we made this nice and consistent. If an app is doing something 'naughty', perhaps we can at least print helpful messages? For example, if you try to render a destroyed mesh, we can console log something informative.

Reply:
Might be better to pass in an optional destruction function (where default implementation calls `.destroy`). And IMO still better to wrap this in https://github.com/playcanvas/engine/pull/2650#issuecomment-754009364

Reply:
Thinking about it more, perhaps refCounting class should never destroy object. Basically keep the engine as it was - refCounting only counts .. but the owner code decides what to do with it.
The common problematic case is that user creates few meshes, and stores them. They have refCount 0. Then the user wants to assign them to MeshInstance for rendering. This increases ref count to 1. When the user then removes it from meshInstance, it would drop to 0 and get destroyed. And so users reference would become invalid.
I think I'll remove auto destroy completely, and change the new user of refCounting (Mprph / MorphInstance) to match Mesh / MeshInstance in a way it destroy Morph as needed instead of depending on it to self-destruct.

Reply:
In which case might be worth checking whether a base class is needed at all.

Reply:
the base class makes all classes to use the same interface. I like that.
When adding it, I considering adding this functionality to each class using mixins (https://javascript.info/mixins) but I find the actual base class more explicit.

Reply:
actually, maybe I'll just use mixin, it's probably suitable for this.